crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -12,48 +12,315 @@
+use tedge_mqtt_ext::Topic;
 use tracing::error;
 
 pub mod config_snapshot;
 pub mod config_update;
 pub mod firmware_update;
 pub mod log_upload;
+mod restart;
+mod software_list;
+mod software_update;
+mod upload;
 
-/// Represents a pending download performed by the downloader from the FTS.
+/// Handles operations.
+///
+/// Handling an operation usually consists of 3 steps:
+///
+/// 1. Receive a smartrest message which is an operation request, convert it to thin-edge message,
+///    and publish on local MQTT (done by the converter).
+/// 2. Various local thin-edge components/services (e.g. tedge-agent) execute the operation, and
+///    when they're done, they publish an MQTT message with 'status: successful/failed'
+/// 3. The cumulocity mapper needs to do some additional steps, like downloading/uploading files via
+///    HTTP, or talking to C8y via HTTP proxy, before it can send operation response via the bridge
+///    and then clear the local MQTT operation topic.
+///
+/// This struct concerns itself with performing step 3.
 ///
-/// Functions which download files from the tedge File Transfer Service as part of handling
-/// operations (e.g. when performing `log_upload` or `config_snapshot`, the relevant file is
-/// uploaded into FTS) will use this type for communicating with the Downloader actor.
-pub struct FtsDownloadOperationData {
-    pub download_type: FtsDownloadOperationType,
-    pub url: String,
+/// Incoming operation-related MQTT messages need to be passed to the [`Self::handle`] method, which
+/// performs an operation in the background in separate tasks. The operation tasks themselves handle
+/// reporting their success/failure.
+pub struct OperationHandler {
+    context: Arc<OperationContext>,
+    running_operations: HashMap<Arc<str>, RunningOperation>,
+}
+
+impl OperationHandler {
+    pub fn new(
+        c8y_mapper_config: &C8yMapperConfig,
+
+        downloader: ClientMessageBox<IdDownloadRequest, IdDownloadResult>,
+        uploader: ClientMessageBox<IdUploadRequest, IdUploadResult>,
+        mqtt_publisher: LoggingSender<MqttMessage>,
+
+        http_proxy: C8YHttpProxy,
+        auth_proxy: ProxyUrlGenerator,
+    ) -> Self {
+        Self {
+            context: Arc::new(OperationContext {
+                capabilities: c8y_mapper_config.capabilities,
+                auto_log_upload: c8y_mapper_config.auto_log_upload,
+                tedge_http_host: c8y_mapper_config.tedge_http_host.clone(),
+                tmp_dir: c8y_mapper_config.tmp_dir.clone(),
+                mqtt_schema: c8y_mapper_config.mqtt_schema.clone(),
+                mqtt_publisher: mqtt_publisher.clone(),
+                software_management_api: c8y_mapper_config.software_management_api,
+
+                // TODO(marcel): would be good not to generate new ids from running operations, see if
+                // we can remove it somehow
+                command_id: IdGenerator::new(crate::converter::REQUESTER_NAME),
```

Comment:
Agree that it would be good to have *no* `IdGenerator` at all outside the `CumulucityConverter`.

This generator is used by the `OperationContext` only in one place: to trigger a software list request after software updates. One should be able to move the call to `request_software_list` from `OperationContext::publish_software_update_status` to the `CumulucityConverter` where the outcome of the software update is forwarded to c8y.